### PR TITLE
Feat/i2c

### PR DESCRIPTION
**Pull request by DhruvaG2000**

### CODE REVIEW

File: libraries/Wire/Wire.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#include "Wire.h"
+
+void arduino::ZephyrI2C::begin() {
+  // const struct device *dev = DEVICE_DT_GET_ANY(adi_adxl372);
+  *i2c_dev = DEVICE_DT_GET(DT_NODELABEL(i2c0));
```
Comment (alvarowolfx) [inline libraries/Wire/Wire.cpp:5]: Needs to create the `bus` with the i2c device + device address, like this https://github.com/zephyrproject-rtos/zephyr/blob/f9d69c34c9d003e99a026ee2a0da9c634e9598c7/drivers/adc/adc_ads1x1x.c#L153
```
struct i2c_dt_spec bus;
```

File: libraries/Wire/Wire.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+void arduino::ZephyrI2C::begin(uint8_t slaveAddr) {
+
+}
+
+void arduino::ZephyrI2C::end() {}
+
+void arduino::ZephyrI2C::setClock(uint32_t freq) {}
+
+void arduino::ZephyrI2C::beginTransmission(uint8_t address) { // TODO for ADS1115
```
Comment (alvarowolfx) [inline libraries/Wire/Wire.cpp:16]: Needs to create the `bus` with the i2c device + device address, like this https://github.com/zephyrproject-rtos/zephyr/blob/f9d69c34c9d003e99a026ee2a0da9c634e9598c7/drivers/adc/adc_ads1x1x.c#L153
```
struct i2c_dt_spec bus;
```

File: libraries/Wire/Wire.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+}
+
+uint8_t arduino::ZephyrI2C::endTransmission(bool stopBit) { return 2; }
+
+uint8_t arduino::ZephyrI2C::endTransmission(void) { // TODO for ADS1115
```
Comment (alvarowolfx) [inline libraries/Wire/Wire.cpp:29]: Call  `i2c_write_dt`

```
bus = i2c_dt_spec{ _addr + i2c device }
ret = i2c_write_dt(bus, _txBuf, sizeof(_txBuf));
```

File: libraries/Wire/Wire.cpp (hunk 2 above)
Comment (alvarowolfx) [reply libraries/Wire/Wire.cpp:16]: sets `_addr` and reset `txBuf`

File: libraries/Wire/Wire.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+size_t arduino::ZephyrI2C::requestFrom(uint8_t address, size_t len) { // TODO for ADS1115
+  return requestFrom(address, len, true);
+}
+
+size_t arduino::ZephyrI2C::write(uint8_t data) {  // TODO for ADS1115 
+  return 1; }
+
+size_t arduino::ZephyrI2C::write(const uint8_t *buffer, size_t size) {
```
Comment (alvarowolfx) [inline libraries/Wire/Wire.cpp:59]: append data to `txBuf` 

```
txBuf[usedTxBuffer++] = data;
```

File: libraries/Wire/Wire.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+uint8_t arduino::ZephyrI2C::endTransmission(bool stopBit) { return 2; }
+
+uint8_t arduino::ZephyrI2C::endTransmission(void) { // TODO for ADS1115
+  return endTransmission(true);
+}
+
+size_t arduino::ZephyrI2C::requestFrom(uint8_t address, size_t len,
+                                       bool stopBit) {
+
+  return len;
```
Comment (alvarowolfx) [inline libraries/Wire/Wire.cpp:47]: Try to use `i2c_burst_read_dt` to read multiple bytes at the same time
Example: https://github.com/zephyrproject-rtos/zephyr/blob/f9d69c34c9d003e99a026ee2a0da9c634e9598c7/drivers/adc/adc_ads1x1x.c#L179

Comment (alvarowolfx) [reply libraries/Wire/Wire.cpp:47]: Zephyr has Ring Buffers too, so we can replicate the MBED implementation 

https://docs.zephyrproject.org/3.0.0/reference/data_structures/ring_buffers.html
MBED: 

* https://github.com/arduino/ArduinoCore-mbed/blob/1da4e665159e7133280c5c33426fb6c22581dc7d/libraries/Wire/Wire.h#L79
* https://github.com/arduino/ArduinoCore-mbed/blob/1da4e665159e7133280c5c33426fb6c22581dc7d/libraries/Wire/Wire.cpp#L119

**Comment by DhruvaG2000** [comment]
planning to add this example:
```C++
#include <Wire.h>

void setup() {
Serial.begin(115200);
Wire.begin();

Wire.beginTransmission(0x53);
Wire.write(0x2C); 
Wire.write(0x08); 
Wire.endTransmission();

Wire.beginTransmission(0x53);
Wire.write(0x31); 
Wire.write(0x08); 
Wire.endTransmission();

Wire.beginTransmission(0x53);
Wire.write(0x2D); 
Wire.write(0x08); 
Wire.endTransmission();
}

void loop() {
Wire.beginTransmission(0x53);
Wire.write(0x32); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte x0 = Wire.read();

Wire.beginTransmission(0x53);
Wire.write(0x33); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte x1 = Wire.read();
x1 = x1 & 0x03;

uint16_t x = (x1 << 8) + x0;
int16_t xf = x;
if(xf > 511)
{
xf = xf - 1024;
}
float xa = xf * 0.004;
Serial.print("X = "); 
Serial.print(xa);
Serial.print(" g"); 
Serial.println(); 


Wire.beginTransmission(0x53);
Wire.write(0x34); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte y0 = Wire.read();

Wire.beginTransmission(0x53);
Wire.write(0x35); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte y1 = Wire.read();
y1 = y1 & 0x03;

uint16_t y = (y1 << 8) + y0;
int16_t yf = y;
if(yf > 511)
{
yf = yf - 1024;
}
float ya = yf * 0.004;
Serial.print("Y = "); 
Serial.print(ya);
Serial.print(" g"); 
Serial.println(); 

Wire.beginTransmission(0x53);
Wire.write(0x36); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte z0 = Wire.read();

Wire.beginTransmission(0x53);
Wire.write(0x37); 
Wire.endTransmission();
Wire.requestFrom(0x53, 1);
byte z1 = Wire.read();
z1 = z1 & 0x03;

uint16_t z = (z1 << 8) + z0;
int16_t zf = z;
if(zf > 511)
{
zf = zf - 1024;
}
float za = zf * 0.004;
Serial.print("Z = "); 
Serial.print(za);
Serial.print(" g"); 
Serial.println(); 
Serial.println(); 
delay(500);
}
```

**Comment by DhruvaG2000** [comment]
Added a much cleaner PR here: https://github.com/zephyrproject-rtos/gsoc-2022-arduino-core/pull/8

Closing this because this repo will not be used in near future.

**Pull request closed without merging**